codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -183,28 +204,45 @@ public void incrementWindowSize(ChannelHandlerContext ctx, Http2Stream stream, i
             // Update the stream window and write any pending frames for the stream.
             AbstractState state = state(stream);
             state.incrementStreamWindow(delta);
-            state.writeBytes(state.writableWindow());
+
+            assert !isWriting;
+            isWriting = true;
+            try {
+                state.writeBytes(writableBytes(state.writableWindow()));
+            } finally {
+                isWriting = false;
+            }
         }
     }
 
     @Override
-    public void sendFlowControlled(ChannelHandlerContext ctx, Http2Stream stream, FlowControlled frame) {
-        checkNotNull(ctx, "ctx");
+    public void sendFlowControlled(Http2Stream stream, FlowControlled frame) {
         checkNotNull(frame, "frame");
-        if (this.ctx != null && this.ctx != ctx) {
-            throw new IllegalArgumentException("Writing data from multiple ChannelHandlerContexts is not supported");
-        }
-        // Save the context. We'll use this later when we write pending bytes.
-        this.ctx = ctx;
         final AbstractState state;
         try {
             state = state(stream);
             state.enqueueFrame(frame);
         } catch (Throwable t) {
-            frame.error(t);
+            frame.error(ctx, t);
             return;
         }
-        state.writeBytes(state.writableWindow());
+
+        assert !isWriting;
+        isWriting = true;
+        try {
+            state.writeBytes(writableBytes(state.writableWindow()));
+        } finally {
+            isWriting = false;
+        }
+    }
+
+    @Override
+    public void writabilityChanged() throws Http2Exception {
```

Comment:
The interface in #3860 could be changed to 

writePendingBytes(int bytesToWrite)

which keep calculations of writable bytes in the handler and out of the flow-controller